src/packet_distribution.cpp

```diff
@@ -455,11 +389,24 @@ void AOApplication::server_packet_received(AOPacket *p_packet)
     if (lobby_constructed)
       w_courtroom->append_ms_chatmessage("", w_lobby->get_chatlog());
 
+    char_mutex.lock();
+    if (char_loader.isRunning())
+        charsLoaded.wait(&char_mutex);
+    char_mutex.unlock();
```

Comment:
use `charsLoaded.waitForFinished()`, but note that both your code and waitForFinished are going to block the main thread which will look like freezing.

so, maybe what you really want is a `QFutureWatcher` where when both futures are done, you then mark the loading process as finished.

however, at second glance, I'm not really seeing any I/O intensive parts in either of those two "loading" functions?